### PR TITLE
Python 3.4/5/6 and PyPy support.

### PR DESCRIPTION
**Pull request by jamadden**
WIP pending zope.app.http fixes and possibly zope.app.exceptions fixes.

I also still haven't checked the coverage numbers very closely.

### CODE REVIEW

**Comment by jamadden** [comment]
The two dependencies have been released and coverage is at 100%. I think this is ready.

File: src/zope/app/file/browser/file.py [hunk 1]
```diff
@@ -293,7 +305,7 @@ def update_object(self, data, contenttype):
         # Update *only* if a new value is specified
         if data:
             self.context.data = data
-            descriptor.attributes += "data",
+            descriptor.attributes += ("data",)
```
Comment (mgedmin) [inline src/zope/app/file/browser/file.py:308]: LOL

File: src/zope/app/file/browser/file.rst [hunk 2]
```diff
@@ -585,13 +572,18 @@ Likewise, the user is not allowed to specify a character set that is not support
...
 
-If you trick Zope and upload a file with a content type that does not match the
-file contents, you will not be able to access the edit view.
+If you trick Zope and upload a file with a content type that does not
+match the file contents, you will not be able to access the edit view.
+(Because of `a bug in zope.app.exception
+<https://github.com/zopefoundation/zope.app.exception/issues/2>`_, we
+can't see the pretty HTML on Python 3 so we resort to looking at the
+exception. When that bug is fixed we should return to
+``handle_errors=True``.)
```
Comment (mgedmin) [inline src/zope/app/file/browser/file.rst]: Wasn't that fixed already?

File: src/zope/app/file/file.py [hunk 3]
```diff
@@ -243,14 +248,20 @@ def __str__(self):
...
             next = self.next
 
-        return ''.join(result)
+        return b''.join(result)
+
+    if str is bytes:
+        __str__ = __bytes__
+    else:
+        def __str__(self):
+            return self.__bytes__().decode("iso-8859-1", errors='ignore')
```
Comment (mgedmin) [inline src/zope/app/file/file.py:257]: `bytestring.decode('ISO-8859-1')` can never raise a UnicodeDecodeError, so you could drop the `errors='ignore'`.  It's harmless, though.

File: src/zope/app/file/image.py [hunk 4]
```diff
@@ -77,86 +78,87 @@ def sizeForDisplay(self):
...
-            while (b and ord(b) != 0xDA):
-                while (ord(b) != 0xFF): b = jpeg.read(1)
-                while (ord(b) == 0xFF): b = jpeg.read(1)
-                if (ord(b) >= 0xC0 and ord(b) <= 0xC3):
+            while b and ord(b) != 0xDA:
+                while ord(b) != 0xFF:
+                    b = jpeg.read(1)
+                while ord(b) == 0xFF:
+                    b = jpeg.read(1)
+                if ord(b) >= 0xC0 and ord(b) <= 0xC3:
```
Comment (mgedmin) [inline src/zope/app/file/image.py]: Could be rewritten to `if 0xC0 <= ord(b) <= 0xC3:`

**Review by mgedmin — APPROVED** [review]
LGTM

**Review by mgedmin — APPROVED** [review]

File: src/zope/app/file/browser/file.py (hunk 1 above)
Comment (jamadden) [reply src/zope/app/file/browser/file.py:308]: pylint recently got smart enough to complain that a trailing comma with no parenthesis can be hard/confusing to read and is as likely to be an accident as it is to be correct (and I have seen that in the wild).

File: src/zope/app/file/browser/file.rst (hunk 2 above)
Comment (jamadden) [reply src/zope/app/file/browser/file.rst]: Yup. I updated the test but not the comment. Thanks.

File: src/zope/app/file/browser/file.py (hunk 1 above)
Comment (mgedmin) [reply src/zope/app/file/browser/file.py:308]: It is confusing -- my LOL was prompted by me not noticing there was a trailing comma in the original, and assuming the old code had a bug.

File: src/zope/app/file/file.py (hunk 3 above)
Comment (jamadden) [reply src/zope/app/file/file.py:257]: Right. I think it saves a little bit of brain-power when reading though, even if you know nothing about codecs.

**Comment by jamadden** [comment]
Thank you! Could we get a PyPI release?

**Comment by mgedmin** [comment]
You have PyPI rights.

I can make a release if you don't want to, but it won't be soon.

**Comment by jamadden** [comment]
Thanks, I'll handle it.

**Pull request merged**